### ISSUE TITLE
maildir.py should conform to docstring and whitespace standards

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6651]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)| @staceysern reported|
|-|-|
|Trac ID|trac#6651|
|Type|enhancement|
|Created|2013-07-28 15:57:56Z|
|Branch|https://github.com/twisted/twisted/tree/maildir-doc-6651|

The docstrings in maildir.py should be updated to provide descriptions of all classes, functions and methods as well as their parameters and attributes. The spacing before classes and functions should conform to the coding standards.

Attachments:

* [maildir-doc-6651-15.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c28/c28ad57b485c69107140c81172eb0542a139ee8d/f5c2a42bda316bcd28a09553842a06ffb3c9245c.patch) (2011 bytes) - added by bbain on 2013-12-02 21:46:49Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6651 6651
type__enhancement enhancement
reporter__shira shira
priority__normal normal
milestone__ 
branch__branches_maildir_doc_6651 branches/maildir-doc-6651
branch_author__shira shira
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__documentation documentation
time__1375027076000000 1375027076000000
changetime__1389668599000000 1389668599000000
version__None None
owner__shira shira

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_1"></a>|@staceysern commented|
|-|-|

(In [39360]) Branching to 'maildir-doc-6651'

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_2"></a>|@staceysern commented|
|-|-|

(In [39384]) Fix docstrings and whitespace.

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_3"></a>|@staceysern commented|
|-|-|

(In [39386]) Add missing docstrings and fix typo.

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_4"></a>|@staceysern commented|
|-|-|

(In [39388]) Remove trailing whitespace.

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_6"></a>|@staceysern commented|
|-|-|

(In [39483]) Added @raises to docstrings for MaildirMailbox and StringListMailbox.

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_7"></a>|@staceysern commented|
|-|-|

(In [39497]) Document one more exception.

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern set owner to @staceysern|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_9"></a>|@staceysern commented|
|-|-|

(In [39781]) Revised API documentation

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_10"></a>|@staceysern removed owner|
|-|-|

Revised to incorporate comments from other mail documentation code reviews and for general improvement in readability. Revisions include:

 * Changing type references from str to bytes
 * Using L{} instead of C{} for standard library types
 * Including a locally usable name in L{}
 * Removing blank lines after class docstrings
 * Removing (optional) in param descriptions
 * Changing the format of type descriptions of tuples
 * Changing the top file message
 * Clarifying that args are parameters for the init function, not the base class itself

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set owner to @staceysern|
|-|-|

Hi Shira, thanks for working on this!

1. Please use L{None} when referring to `None` in the docstrings - it distinguishes that it returns an instance of NoneType that is `None`, and not the string "None". Instances of this are on lines 200 and 811.

2. Missing a full stop on lines 130, 133 and 184.

3. A few methods in maildir.AbstractMaildirDomain say they'll do something, but will fail/throw an exception if not overridden. I think these should have notes saying that subclasses must implement this behaviour in the docstring.

4. In MaildirDirdbmDomain, postmaster is a documented as a boolean, but the default value is an int. It seems to treat it as a boolean anyway (line 814), so this might not be an issue. I can't see it being used anywhere else, so I don't think it is.

Please fix/address these and resubmit. :)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_13"></a>|@staceysern commented|
|-|-|

(In [40055]) Incorporate review comments

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_14"></a>|@staceysern removed owner|
|-|-|

For item 1, I changed None to C{None} (not L{None}).  None is a Python constant like True and False (actually it is more of a constant since you can't change the value unlike True and False).  As an aside, L{None} links to something totally unrelated to None. 

I fixed items 2 and 3.

For item 4, the default value is an int but the variable is treated like a boolean.  Since booleans are integer values in Python (True is 1 and False is 0), I thought it made the most sense to describe the variable as a boolean.

**Comment by twisted-trac** [comment]
|<img alt="bbain's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|bbain commented|
|-|-|

Removed escaping and fixed indexing of tuples.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj set owner to @staceysern<br>@wallrj set status to `new`|
|-|-|

Thanks shira,

Branch looks great...as usual. Looking forward to seeing all your documentation
branches land in trunk.

Here are a few notes and nitpicks.

Notes:
 1. Merges cleanly
 2. Decided not to apply maildir-doc-6651-15.patch from bbain (I don't agree
    with the change)
 3. Clean build results
    1. https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/maildir-doc-6651

Points:
 1. [branches/maildir-doc-6651/twisted/mail/maildir.py](branches/maildir-doc-6651/twisted/mail/maildir.py)
    1. "55          @ivar _clock: See L{__init__} " Maybe say "See C{clock} of L{__init__}" as mentioned in the coding standard.
       1. https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto9
    2. "2-L{tuple} of (E{1}) L{bytes}, (E{2}) L{bytes}" I mentioned it in other
       reviews, and I prefer the syntax suggested by exarkun in r41111
       1. Same elsewhere
    3. "L{dict} of  L{bytes} -> L{AliasBase}" maybe would be clearer as "L{dict} mapping L{bytes} to L{AliasBase}"
       1. Same elsewhere
    4. "L{NoneType <types.NoneType>}" to L{None}
    5. "C{None}" to "L{None}"
       1. Repeated later in the branch
    6. "@type user: L{Address}" I looked at this and was confused for a moment,
       because the "user" arguments of the surrounding methods represent a user
       name. I guess in this case it's an email address object. Bad variable
       name I suppose. Nothing you can do about that here.
    7. "@rtype: L{list} of L{ICredentialsChecker <checkers.ICredentialsChecker>} provider" should be "providers" (plural) I think.
    8. "A deferred which triggers" I think we say "A deferred which fires"
       elsewhere in the codebase. Or at least "A deferred which is triggered"
    9. Typo "the new/ directory"
       1. Or maybe it isn't and should be I{new/}
       2. Same for ".Trash/"
       3. Same elsewhere in the branch
    10. "if none is specified" maybe use L{None}
        1. Same elsewhere in the branch
    11. "DirDBM" might be nice to put this in I{} or L{} or even make it a link
        to an online reference page.
    12. "n-L{tuple}" not sure if the "n-" is really necessary

Please merge after addressing (or answering) the numbered points above. I guess
there may be some other style changes which have been suggested in the code
reviews for your other branches. So you'll probably want to bring this branch
into line with those.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_18"></a>|@staceysern commented|
|-|-|

(In [41229]) Incorporate code review comments.

refs [#6651](https://github.com/twisted/twisted/issues/6651)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_19"></a>|@staceysern set status to `closed`|
|-|-|

(In [41235]) Merge maildir-doc-6651: Fix docstrings in maildir.py

Author: shira
Reviewer: hawkowl, rwall
Fixes: [#6651](https://github.com/twisted/twisted/issues/6651)
